Add Base.getindex() support

samoconnor opened the pull request:
Adds support for Indexable syntax, e.g. `xml[tag1][tag2][tag3]`. See full examples in README.md.

Also adds convenience constructors:
```julia
parse_string(s::Array{UInt8,1}) = parse_string(bytestring(s))
XMLDocument(xml) = parse_string(xml)
XML(xml) = XMLDocument(xml)
``` 

The aim is to have a quick way to extract particular tags from web service responses like this:

```julia
xml = """
<?xml version="1.0" encoding="UTF-8"?>
<ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01">
  <Owner>
    <ID>bcaf1ffd86f461ca5fb16fd081034f</ID>
    <DisplayName>webfile</DisplayName>
  </Owner>
  <Buckets>
    <Bucket>
      <Name>quotes</Name>
      <CreationDate>2006-02-03T16:45:09.000Z</CreationDate>
    </Bucket>
    <Bucket>
      <Name>samples</Name>
      <CreationDate>2006-02-03T16:41:58.000Z</CreationDate>
    </Bucket>
  </Buckets>
</ListAllMyBucketsResult>
"""

@test XML(xml)["Buckets"]["Bucket"]["Name"] == ["quotes", "samples"]
```

tkelman commented:
I don't think this is really appropriate for the multiple return case. Particularly getindex of an array of XMLElements mapping the getindex to its children, doesn't feel right. xpath is the well established standard semantics for doing this king of thing, I'd rather not invent a Julia-specific method of indexing xml.

nalimilan commented:
+1 for the `getindex` syntax, but as @tkelman said indexing an array of `XMLElement` shouldn't behave differently than indexing any array. One should use `map` to index all elements of the array.

nalimilan commented on src/getindex.jl:
```diff
@@ -0,0 +1,28 @@
+# Indexable Collection support
+
+function Base.getindex(e::XMLElement, tag::AbstractString)
+
```
No space at the beginning of a function (here an elsewhere).

nalimilan commented on src/getindex.jl:
```diff
@@ -0,0 +1,28 @@
+end
+
+
+Base.getindex(d::XMLDocument, tag::AbstractString) = LightXML.root(d)[tag]
```
You probably don't need to write `LightXML.` here.

nalimilan commented on src/nodes.jl:
```diff
@@ -266,6 +266,9 @@ Base.next(it::XMLElementIter, p::Xptr) = (XMLElement(p), ccall((:xmlNextElementS
 
 child_elements(x::XMLElement) = XMLElementIter(x.node.ptr)
 
+has_child_elements(x::XMLElement) = !isempty(child_elements(x))
```
Given how simple and readable is `!isempty(child_elements(x))`, I'm not sure a function is needed here, especially since it's used in a single place.

tkelman commented:
`getindex` syntax would probably work well as a shorthand for xpath queries, ref #38. LibExpat.jl does things that way and works pretty well. It would also likely be quite a bit better performance than iterating through all children.

I don't particularly agree with the shorthands being added here - `XML` is too generic to export, `XMLDocument` can be created from either a file or a string and I don't think it's clear-cut that one should be given preferential treatment with constructor syntax, and an `Array{UInt8,1}` isn't a string - if that's what your user code is dealing with then it would be more explicit to call `bytestring` at the call site, not inside this package.

samoconnor commented:
@tkelman, @nalimilan, thanks for the feedback.

> getindex of an array of XMLElements mapping the getindex to its children, doesn't feel right

I agree, I've committed a change that adds a new type to avoid this.

> I don't particularly agree with the shorthands being added here

Removed.

> Re: has_child_elements()

Removed.

>  You probably don't need to write LightXML. here

Removed.

samoconnor commented:
Re: @tkelman's comment *"unlike JSON, XML isn't an associative array abstraction"*

I can agree with that to some extent in the general case. XML can be used in many ways.   
However, in the systems I'm working with XML is used heavily as an associative array abstraction for web service responses. XML is what was before everyone realised that JSON does it better. I hope XML goes away completely for this sort of thing, but that will take time. For now I would love to be able to index into an XML response with exactly the same syntax I use to index into a JSON response.

Re: @tkelman's comment *"it [xpath] would also likely be quite a bit better performance than iterating through all children."*

I'm all for ways to go faster, so...
I've merged #38 into my local branch and done some performance tests based on the test cases in #39.

The full test code is below, the result is that xpath does not seem to be faster:

```
test A getindex  1.15 seconds 
test A xpath     2.01 seconds 

test B getindex  5.97 seconds 
test B xpath     7.76 seconds

test C getindex  1.96 seconds 
test C xpath     3.09 seconds 
```


```julia
xml = """
<CreateQueueResponse>
    <CreateQueueResult>
        <QueueUrl>
            http://queue.amazonaws.com/123456789012/testQueue
        </QueueUrl>
    </CreateQueueResult>
    <ResponseMetadata>
        <RequestId>
            7a62c49f-347e-4fc4-9331-6e8e7a96aa73
        </RequestId>
    </ResponseMetadata>
</CreateQueueResponse>
"""

function f()
    x = parse_string(xml)
    @test x["CreateQueueResult"]["QueueUrl"] ==
          "http://queue.amazonaws.com/123456789012/testQueue"
end

f() 
println("test A getindex:")
@time for i = 1:100000 f() end


function f()
    x = parse_string(xml)
    @test strip(content(evalxpath("//CreateQueueResult/QueueUrl", x)[1])) ==
          "http://queue.amazonaws.com/123456789012/testQueue"
end
f() 
println("test A xpath:")
@time for i = 1:100000 f() end

xml = """
<GetQueueAttributesResponse>
  <GetQueueAttributesResult>
    <Attribute>
      <Name>ReceiveMessageWaitTimeSeconds</Name>
      <Value>2</Value>
    </Attribute>
    <Attribute>
      <Name>VisibilityTimeout</Name>
      <Value>30</Value>
    </Attribute>
    <Attribute>
      <Name>ApproximateNumberOfMessages</Name>
      <Value>0</Value>
    </Attribute>
    <Attribute>
      <Name>ApproximateNumberOfMessagesNotVisible</Name>
      <Value>0</Value>
    </Attribute>
    <Attribute>
      <Name>CreatedTimestamp</Name>
      <Value>1286771522</Value>
    </Attribute>
    <Attribute>
      <Name>LastModifiedTimestamp</Name>
      <Value>1286771522</Value>
    </Attribute>
    <Attribute>
      <Name>QueueArn</Name>
      <Value>arn:aws:sqs:us-east-1:123456789012:qfoo</Value>
    </Attribute>
    <Attribute>
      <Name>MaximumMessageSize</Name>
      <Value>8192</Value>
    </Attribute>
    <Attribute>
      <Name>MessageRetentionPeriod</Name>
      <Value>345600</Value>
    </Attribute>
  </GetQueueAttributesResult>
  <ResponseMetadata>
    <RequestId>1ea71be5-b5a2-4f9d-b85a-945d8d08cd0b</RequestId>
  </ResponseMetadata>
</GetQueueAttributesResponse>
"""

function f()
    x = parse_string(xml)
    d = x["GetQueueAttributesResult"]["Attribute"]["Name", "Value"]
    @test d["MessageRetentionPeriod"] == "345600"
    @test d["CreatedTimestamp"] == "1286771522"
end
f()
println("test B getindex")
@time for i = 1:100000 f() end

function f()
    x = parse_string(xml)
    r = evalxpath("//GetQueueAttributesResult/Attribute/*[self::Name or self::Value]", x)
    d = [strip(content(r[i])) => strip(content(r[i+1])) for i = 1:2:length(r)]
    @test d["MessageRetentionPeriod"] == "345600"
    @test d["CreatedTimestamp"] == "1286771522"
end
f()
println("test B xpath")
@time for i = 1:100000 f() end

xml = """
<?xml version="1.0" encoding="UTF-8"?>
<ListAllMyBucketsResult>
  <Owner>
    <ID>bcaf1ffd86f461ca5fb16fd081034f</ID>
    <DisplayName>webfile</DisplayName>
  </Owner>
  <Buckets>
    <Bucket>
      <Name>quotes</Name>
      <CreationDate>2006-02-03T16:45:09.000Z</CreationDate>
    </Bucket>
    <Bucket>
      <Name>samples</Name>
      <CreationDate>2006-02-03T16:41:58.000Z</CreationDate>
    </Bucket>
  </Buckets>
</ListAllMyBucketsResult>
"""

function f()
    @test parse_string(xml)["Buckets"]["Bucket"]["Name"] == ["quotes", "samples"]
end
f()
println("test C getindex")
@time for i = 1:100000 f() end

function f()
    x = parse_string(xml)
    r = evalxpath("//Buckets/Bucket/Name", x)
    @test [strip(content(i)) for i in r] == ["quotes", "samples"]
end
f()
println("test C xpath")
@time for i = 1:100000 f() end

tkelman commented:
That's surprising, but good to quantify. I suppose we could proceed with a very reduced version of this, but it's going to need to remove a lot of pieces that are semantically questionable. The "multidimensional" indexing that returns a dict makes no sense to me, is there any precedebt at all for that kind of indexing? If not, let's leave it out.

Don't make a new type for `Array{XMLElement,1}` - there's no need for a wrapper type, the standard Julia array type will do just fine. Indexing an array-like object, whether it has scalars or strings or `XMLElement`s in it, is indexing with respect to that outer array, and should not propagate as a vectorizedindexing operation on the elements in the array. Use `map` or a comprehension for that.

The basic indexing of a tag should always return an array of `XMLElement`s for type stability and consistency. That array can be length 1 for a single match, or length 0 for no matches.

tkelman commented:
Also, can't you ask for content-type=json from most web api's these days?

samoconnor commented:
Asking for Content-type: json is a really good tip!!! Thanks @tkelman 

A bit of googling suggests that some AWS services accept `ContentType=JSON` as an (undocumented) query parameter. I've got this working for SQS, SNS and IAM. Good riddance to XML for those services.

It seems that S3 and SDB both ignore the ContentType=JSON parameter.
I've also tried putting `Accept: application/json` or `Content-Type: application/json` (long shot) in the request header. No Luck. So, still stuck with XML for now...

samoconnor commented:
I've removed the indexing of the collection of elements by tag per @tkelman's request.
A comprehension now achieves the same thing (albeit with a little more syntactic noise):

```julia
Old: xml["GetQueueAttributesResult"]["Attribute"]["Name" => "Value"]
New: [a["Name"] => a["Value"] for a in xml["GetQueueAttributesResult"]["Attribute"]]

Old: xml["Buckets"]["Bucket"]["Bucket"]
New: [b["Name"] for b in xml["Buckets"]["Bucket"]]
```

I'm afraid I don't buy the argument that says `getindex()` should always return `Array{XMLElement,1}` because type consistency is a "good thing". Why have a dynamic language if you're not allowed to have dynamic return types where it is convenient?

For performance critical code, sure, by type stable. But I've already demonstrated that this code is faster than the "proper" xpath way. In this context, next to the overhead of parsing XML, a little dynamic dispatch overhead is insignificant. The golden rule of performance optimisation is that it should be baed on measurement.

tkelman commented:
Type stability is absolutely crucial to performance in Julia code. If you want dynamic return types and none of the benefits of type inference, there are lots of interpreted languages out there. Code that is not amenable to type inference is technical debt that ends up forcing someone else to rewrite it later.

samoconnor commented:
@tkelman, can you propose a real-world use case of `getindex(x::XMLElement,tag)` where performance would be impacted by the dynamic return type?
I'm not happy with changing code for "performance" reasons without making measurements. But I am happy to code up a test framework to simulate a real-world use case where performance matters.

Looking at it another way, is the following JSON example with dynamic return type ok? or is it a "technical debt that ends up forcing someone else to rewrite it later" ...

```julia
a = JSON.parse("""{
  "Statement": {
    "Effect": "Allow",
    "Action": "s3:ListBucket",
    "Resource": "arn:aws:s3:::example_bucket"
  }
}""")

b = JSON.parse("""{
  "Statement": {
    "Effect": "Allow",
    "Action": "s3:*",
    "Resource": [
      "arn:aws:s3:::mybucket",
      "arn:aws:s3:::mybucket/*"
    ]
  }
}""")

@test a["Statement"]["Resource"] == "arn:aws:s3:::example_bucket"

@test b["Statement"]["Resource"] == ["arn:aws:s3:::mybucket","arn:aws:s3:::mybucket/*"]
```

tkelman commented:
XML and JSON are not the same, and just can't support exactly the same API - JSON is an associative array so can be converted to a Julia Dict. Heterogenously typed Dicts are not good to base performance critical infrastructure on, just look at all the design flaws in DataFrames that have needed to be redesigned for this reason.

XML is not an associative array and cannot in general be treated as if it's a Julia Dict. I can try to write an example of getindex on a small xml snippet where the ability to infer the result type will matter. It's also just idiomatic Julian API design, type stability is one of the most important API design criteria throughout the standard library. If it's possible to accomplish, then there's no reason to choose an unstable API design over a type stable one. Choosing "convenience" (and I disagree there, I think dealing with different types depending on the number of matches is less convenient, not more) over the ability for the compiler to reason about code is fine in slow interpreted languages, but not in Julia.

nalimilan commented:
Indeed type stability is a must. It could be interesting to have an alternative syntax for when you only expect to get one element. That would allow raising an exception if several elements match, instead of taking the first one without checking. We could imagine using a different type of brackets -- maybe `{}` if they are free at some point.

tkelman commented:
There's too much unrelated stuff in here now. Don't open PR's from master of your fork, make specific branches for individual changes.

samoconnor commented:
> XML is not an associative array and cannot in general be treated as if it's a Julia Dict.

Sure it can, see https://github.com/samoconnor/LightXML.jl/commit/da81f6c7a0f7657103c605646866e8f5a6572f4c .
This includes test cases that verify round-trip conversion xml -> dict -> xml by comparing `xmllint` output.

>  I can try to write an example of getindex on a small xml snippet where the ability to infer the result type will matter.

Please do...

tkelman commented:
The important part of that sentence was "in general" - you've made multiple tags with the same name equivalent to a julia array (and between symbols and attributes, which you did not desvcribe anywhere in the comments here and I don't agree with), which is not an unambiguous conversion. That code has a number of fundamental flaws. It should not be using string concatenation. You're not adding xml tags between child elements of arrays, just appending them on after another with nothing between them.

tkelman commented:
Your getindex also throws away attribute data
```
  | | |_| | | | (_| |  |  Version 0.4.2 (2015-12-06 21:47 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org release
|__/                   |  x86_64-linux-gnu

julia> using LightXML                                                        
julia> Pkg.status("LightXML")
 - LightXML                      0.2.1+             072b1c92

julia> ex = root(parse_string("<ex><el id='0'>0</el><el id='1'>1</el></ex>"))
<ex>
  <el id="0">0</el>
  <el id="1">1</el>
</ex>


julia> LightXML.get_elements_by_tagname(ex, "el")                            2-element Array{LightXML.XMLElement,1}:
 <el id="0">0</el>

 <el id="1">1</el>


julia> ex["el"]
2-element Array{ByteString,1}:
 "0"
 "1"

julia> Pkg.clone("https://github.com/JohnMylesWhite/Benchmarks.jl")
INFO: Cloning Benchmarks from https://github.com/JohnMylesWhite/Benchmarks.jlINFO: Computing changes...

julia> function f1()   
       ex = root(parse_string("<ex><el id='0'>0</el><el id='1'>1</el></ex>"))
       for i=1:10000
           LightXML.get_elements_by_tagname(ex, "el")
       end
       end
f1 (generic function with 1 method)

julia> using Benchmarks
julia> @benchmark f1()
================ Benchmark Results ========================
     Time per evaluation: 4.08 ms [2.96 ms, 5.20 ms]
Proportion of time in GC: 15.50% [0.47%, 30.52%]
        Memory allocated: 5.95 mb
   Number of allocations: 100003 allocations
       Number of samples: 100
   Number of evaluations: 100
 Time spent benchmarking: 0.42 s


julia> function f2()
       ex = root(parse_string("<ex><el id='0'>0</el><el id='1'>1</el></ex>"))
       for i=1:10000
           ex["el"]
       end
       end
f2 (generic function with 1 method)

julia> @benchmark f2()
================ Benchmark Results ========================                       Time per evaluation: 12.27 ms [9.41 ms, 15.13 ms]
Proportion of time in GC: 14.75% [6.96%, 22.53%]
        Memory allocated: 12.51 mb
   Number of allocations: 250003 allocations
       Number of samples: 100
   Number of evaluations: 100
 Time spent benchmarking: 1.25 s
```

tkelman commented:
Do check the `@code_warntype` too, it's quite clear here.

samoconnor commented:
>The important part of that sentence was "in general" - you've made multiple tags with the same name equivalent to a julia array which is not an unambiguous conversion. 

Thinking about this a bit more, I realised that [collecting sub-elements with same tag into a vector](https://github.com/samoconnor/XMLDict.jl/blob/master/src/XMLDict.jl#L157) can alter the order in the case where tag types are interspersed. This is fixed in https://github.com/samoconnor/XMLDict.jl/commit/7daea360ab705a609ff8ce96ca0de058220b0423


I don't think the XML -> Dict transform is ambiguous given that the reverse transform tests all pass:
```julia
@test normalise_xml(xml) == normalise_xml(XMLDict.dict_xml(xml_dict(xml)))
```
These tests include a number of small real-world AWS XML responses and also [the complete XML standard](https://github.com/samoconnor/XMLDict.jl/blob/master/test/REC-xml-20081126.xml) as a more complex XML example. 


> That code has a number of fundamental flaws. It should not be using string concatenation. You're not adding xml tags between child elements of arrays (if strings), just appending them one after another with nothing between them.

Can you suggest a test-case XML document where you think "xml tags between child elements" youldn't work?

samoconnor commented:
> Your getindex also throws away attribute data

Fixed in https://github.com/samoconnor/XMLDict.jl/commit/94717e34bd5086799daf841756574d7f17c6e813

```julia
using XMLDict
d = xml_dict("<ex><el>0</el><el>1</el></ex>")
@test d["ex"]["el"] == ["0", "1"]

d = xml_dict("<ex><el id='0'>0</el><el id='1'>1</el></ex>")
@test d["ex"]["el"] == [Dict(:id=>"0",""=>"0"), Dict(:id=>"1",""=>"1")]
@test d["ex"]["el"][1][:id] == "0"

x = parse_xml("<ex><el id='0'>0</el><el id='1'>1</el></ex>")
@test x["el"][1][""] == "0"
@test x["el"][1][:id] == "0"
```

samoconnor commented:
Re the benchmark...
```julia
for i=1:10000
    LightXML.get_elements_by_tagname(ex, "el")
end
Time per evaluation: 4.08 ms 

for i=1:10000
    ex["el"]
end
Time per evaluation: 12.27 ms 
```

It `ex[]` is 3 x slower, which is not surprising given that it calls `get_elements_by_tagname()` and also `child_elements()` and `child_elements()` and `content()` and `length()` and `strip()`. This just looks like: doing more takes longer.

I fail to see how this measurement has anything to do with return type stability. The return type is not even used in this code.

If you believe that return type stability is performance-critical to `getindex(::XMLElement)`, then show me a benchmark where:
 * a type-stable version of `getindex(::XMLElement)` is faster than the current implementation, and
 * the speed-up is significant in comparison to the rest of a typical XML use-case workload. (i.e. there is no point making `getindex()` 100 times faster if `getindex()` is only 1% if the workload anyway).

tkelman commented:
Yes, the benchmark isn't very good. My proposal for what to do with getindex is just make it shorthand for `get_elements_by_tagname`. I don't have enough time to convince you that type stability is the single most important criteria in Julia API design - there are a number of blog posts by Dahua Lin and John Myles White on the issue.

tkelman commented:
a slightly better benchmark
```
julia> function f2()
       ex = root(parse_string("<ex><el>0</el><el>1</el></ex>"))                     ret = 0
       for i=1:10000
           for el in ex["el"]
               ret += parse(Int, el)
           end
       end
       return ret
       end
f2 (generic function with 1 method)

julia> @benchmark f2()
================ Benchmark Results ========================
     Time per evaluation: 17.83 ms [17.05 ms, 18.60 ms]
Proportion of time in GC: 6.93% [3.50%, 10.37%]
        Memory allocated: 13.12 mb
   Number of allocations: 270003 allocations
       Number of samples: 100
   Number of evaluations: 100
 Time spent benchmarking: 1.81 s


julia> function f1()
       ex = root(parse_string("<ex><el>0</el><el>1</el></ex>"))                     ret = 0
       for i=1:10000
           for el in LightXML.get_elements_by_tagname(ex, "el")
               ret += parse(Int, content(el))
           end
       end
       return ret
       end
f1 (generic function with 1 method)

julia> @benchmark f1()
================ Benchmark Results ========================
     Time per evaluation: 8.90 ms [8.25 ms, 9.54 ms]
Proportion of time in GC: 8.20% [1.46%, 14.95%]
        Memory allocated: 7.78 mb
   Number of allocations: 140003 allocations
       Number of samples: 100
   Number of evaluations: 100
 Time spent benchmarking: 0.91
```

Adding attributes to the data changes the return type of your current getindex, meaning you have to add `content` or not to deal with data that has or does not have attributes.

tkelman commented:
`dict_xml` issues
```
julia> XMLDict.dict_xml(Dict("a"=>Any["b",1]))
ERROR: MethodError: `value_xml` has no method matching value_xml(::Int64)
 in node_xml at /home/tkelman/.julia/v0.4/XMLDict/src/XMLDict.jl:230
 in node_xml at /home/tkelman/.julia/v0.4/XMLDict/src/XMLDict.jl:222
 in value_xml at /home/tkelman/.julia/v0.4/XMLDict/src/XMLDict.jl:235
 in value_xml at /home/tkelman/.julia/v0.4/XMLDict/src/XMLDict.jl:237
 in node_xml at /home/tkelman/.julia/v0.4/XMLDict/src/XMLDict.jl:225
 in node_xml at /home/tkelman/.julia/v0.4/XMLDict/src/XMLDict.jl:222
 in dict_xml at /home/tkelman/.julia/v0.4/XMLDict/src/XMLDict.jl:208

julia> XMLDict.dict_xml(Dict("a"=>Any["b",["c","d"]]))
"<?xml?>\n<a>b</a><a>c</a><a>d</a>"
```

samoconnor commented:
> dict_xml issues ...

It should be no surprise that throwing odd-ball data at a function that expects a certain structure will cause oddness...

```julia
d = Dict(); d[d] = d ; json(d)
ERROR: StackOverflowError:

split(UTF8String(open("/dev/random", "r") do io read(io, UInt8, 10) end))
ERROR: UnicodeError: invalid character index
```

What did you expect?

Please note that neither the [documentation for XMLDict](https://github.com/samoconnor/XMLDict.jl) or the [`exports` statement in XMLDict.jl](https://github.com/samoconnor/XMLDict.jl/blob/master/src/XMLDict.jl#L10) mentions `dict_xml()` at all. The [source comment for `dict_xml()`](https://github.com/samoconnor/XMLDict.jl/blob/master/src/XMLDict.jl#L201) explicitly says that it expects a `Dict` *"produced by xml_dict"*.

`dict_xml()` is primarily intended as a proof that `xml_dict()` is robust and as a test tool.
I don't imagine that there would be much call for producing an XML Dict by hand and then passing it to `dict_xml()`.

It might be useful in do something like this (maybe a good reason to document and export):

```julia
d = xml_dict(get("settings.xml"))
d["user"]["sam"]["homepage"] = "http://octe.ch"
put("settings.xml", dict_xml(d)
```

tkelman commented:
That comment wasn't in the implementation of `dict_xml` that was in this PR. The `Expand heterogeneous array...` lines are the ones I considered likely to be buggy.

Anyway you can go ahead and do XML dicts with slow string concatenation if you want in a package, but only if using that package does not change the behavior of unrelated code that uses LightXML. Don't define Base methods for LightXML's types, that's not kosher. Define wrapper types within XMLDict, then you can do whatever you want with types that your package controls.

samoconnor commented:
Here is a benchmark that tries to address the issue at hand, the type-stability of getindex().

There are two versions of getindex(), one always returns `Array{XMLElement,1}`, the other returns some mess like: `Union{Void,ASCIIString,UTF8String,XMLElement,Array{XMLElement,1}, Array{ASCIIString,1}{Array{UTF8String,1}}`.

The results are below, but in summary both versions take 8 seconds +/- noise to run.

No measurable difference.

```julia
foo.jl:
using LightXML

const type_stable_or_die = true

if type_stable_or_die

function Base.getindex(x::XMLElement, tag::AbstractString)

    return get_elements_by_tagname(x, tag)
end

        
function f(d::XMLDocument)

    t = d["GetUserResult"][1]["User"][1]["PasswordLastUsed"][1]
    t = strip(content(t))
    t = DateTime(t, "yyyy-mm-ddTHH:MM:SSZ")
    t = Dates.datetime2unix(t)
    return t
end

else

function Base.getindex(x::XMLElement, tag::AbstractString)

    if tag == ""
        return strip(content(x))
    end

    l = get_elements_by_tagname(x, tag)
    if isempty(l)
        return nothing
    end
    if isempty(child_elements(l[1])) &&
       isempty(attributes(l[1]))
        l = [strip(content(i)) for i in l]
    end
    return length(l) == 1 ? l[1] : l
end

function f(d::XMLDocument)
    t = d["GetUserResult"]["User"]["PasswordLastUsed"]
    t = DateTime(t, "yyyy-mm-ddTHH:MM:SSZ")
    t = Dates.datetime2unix(t)
    return t
end

end

Base.getindex(x::XMLDocument, tag::AbstractString) = root(x)[tag]

x = LightXML.parse_string("""
<GetUserResponse>
  <GetUserResult>
    <User>
      <PasswordLastUsed>
            2015-12-23T22:45:36Z
      </PasswordLastUsed>
      <Arn>arn:aws:iam::012541411202:root</Arn>
      <UserId>012541411202</UserId>
      <CreateDate>2015-09-15T01:07:23Z</CreateDate>
    </User>
  </GetUserResult>
  <ResponseMetadata>
    <RequestId>837446c9-abaf-11e5-9f63-65ae4344bd73</RequestId>
  </ResponseMetadata>
</GetUserResponse>
""")

function dotest(x)

    count = 0

    for i=1:100000
        count += f(x)
    end
    println(count);
end

println(type_stable_or_die ? "" : "not", " type stable")
@time dotest(x)
```


```
$ julia foo.jl
not type stable
1.450910736e14
  7.962570 seconds (30.56 M allocations: 1.601 GB, 3.88% gc time)
$ julia foo.jl
not type stable
1.450910736e14
  8.036092 seconds (30.56 M allocations: 1.601 GB, 4.08% gc time)
$ julia foo.jl
not type stable
1.450910736e14
  8.088126 seconds (30.56 M allocations: 1.601 GB, 3.96% gc time)
$ julia foo.jl
 type stable
1.450910736e14
  7.903482 seconds (30.44 M allocations: 1.592 GB, 3.91% gc time)
$ julia foo.jl
 type stable
1.450910736e14
  8.034649 seconds (30.44 M allocations: 1.592 GB, 4.11% gc time)
$ julia foo.jl
 type stable
1.450910736e14
  8.386510 seconds (30.44 M allocations: 1.592 GB, 4.13% gc time)
```

tkelman commented:
Profile that benchmark, it spends 95% of its time in dates. Of course if the expensive part of what you're doing isn't navigating the xml tree or dependent on type inference of the results of doing so, you won't care about getindex performance. I'm not going to make any claims or rule out decent performance for use cases where navigating the xml tree is the expensive part.

samoconnor commented:
> Profile that benchmark, it spends 95% of its time in dates. Of course if the expensive part of what you're doing isn't navigating the xml tree or dependent on type inference of the results of doing so, you won't care about getindex performance.

Exactly !!!!

In what real world use case are you ever going to process XML in a way where the XML tree walking part is expensive compared to what you're doing with the data. Parsing a date to an int is well towards the low end of the complexity scale.

If you propose a real world use case where "decent performance" depends on type inference of the `get index()` result, then I will take your concern seriously.

samoconnor commented:
When I replace the date time parsing with a simple `length(string)` to produce the returned int, the timing is as follows.

median type stable  0.28s,
median not type stable 0.30s

7%-ish difference.

```
$ julia foo.jl
 type stable
2000000
  0.253697 seconds (4.35 M allocations: 260.184 MB, 18.26% gc time)
$ julia foo.jl
 type stable
2000000
  0.280087 seconds (4.35 M allocations: 260.184 MB, 22.76% gc time)
$ julia foo.jl
 type stable
2000000
  0.302786 seconds (4.35 M allocations: 260.184 MB, 18.76% gc time)
$ julia foo.jl
not type stable
2000000
  0.325423 seconds (4.56 M allocations: 269.888 MB, 21.09% gc time)
$ julia foo.jl
not type stable
2000000
  0.300404 seconds (4.56 M allocations: 269.888 MB, 16.03% gc time)
$ julia foo.jl
not type stable
2000000
  0.298291 seconds (4.56 M allocations: 269.888 MB, 16.27% gc time)
```

tkelman commented:
IO dominated navigation of deeply nested complicated structures. I do this in CoinOptServices.jl, and on small problems the non-XML-IO part is a few milliseconds, so I'd rather not add latency. Besides, our opinions about which API is cleaner, the type-stable one or the unpredictable messy union one, are clearly not going to get reconciled here. I'd have to code around your API giving me different types on different input data, more downstream work for zero benefit.

tkelman commented:
And post compilation I see 157 ms vs 180 ms mean.

nalimilan commented:
@samoconnor Type-stability is a golden rule in Julia programming, period. In you own code, you can do what you want, but in libraries you cannot know in advance what the bottleneck will be in your user's code. You're wasting your energy if you want to go against this.
